### PR TITLE
GUVNOR-3236: Rearrange menus structure

### PR DESCRIPTION
**Pull request by paulovmr**

### CODE REVIEW

**Comment by paulovmr** [comment]
Part of an ensemble:
* https://github.com/AppFormer/uberfire/pull/798
* https://github.com/kiegroup/kie-wb-common/pull/1053
* https://github.com/kiegroup/drools-wb/pull/572
* https://github.com/kiegroup/kie-wb-distributions/pull/578

File: uberfire-workbench/uberfire-workbench-client-views-patternfly/src/main/java/org/uberfire/client/views/pfly/menu/WorkbenchMenuBarView.java [hunk 1]
```diff
@@ -50,7 +52,8 @@
  * The Menu Bar widget
  */
 @ApplicationScoped
-public class WorkbenchMenuBarView extends Composite implements WorkbenchMenuBarPresenter.View {
+public class WorkbenchMenuBarView extends Composite implements IsWidget,
```
Comment (ederign) [inline uberfire-workbench/uberfire-workbench-client-views-patternfly/src/main/java/org/uberfire/client/views/pfly/menu/WorkbenchMenuBarView.java]: @paulovmr I think now it's a good time to remove Composite/Widget interface from here and use just Errai UI.

File: uberfire-workbench/uberfire-workbench-client-views-patternfly/src/main/java/org/uberfire/client/views/pfly/menu/WorkbenchMenuCompactNavBarView.java [hunk 2]
```diff
@@ -91,6 +92,12 @@ public void addCustomMenuItem(final Widget menu,
     }
 
     @Override
+    public void addCustomMenuItem(final IsElement menu,
+                                  final MenuPosition position) {
+        //No support for adding custom menus when using compact mode.
```
Comment (ederign) [inline uberfire-workbench/uberfire-workbench-client-views-patternfly/src/main/java/org/uberfire/client/views/pfly/menu/WorkbenchMenuCompactNavBarView.java]: Warning on log?

File: uberfire-workbench/uberfire-workbench-client-views-patternfly/src/main/java/org/uberfire/client/views/pfly/menu/megamenu/WorkbenchMegaMenuView.java [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+
+    @PostConstruct
+    public void setup() {
+        final String defaultMenuText = translationService.format(Constants.Menu);
+
+        if (megaMenuBrands.isAmbiguous()) {
+            throw new RuntimeException("Multiple implementations of MegaMenuBrand were provided.");
+        }
+
+        if (!megaMenuBrands.isUnsatisfied()) {
```
Comment (ederign) [inline uberfire-workbench/uberfire-workbench-client-views-patternfly/src/main/java/org/uberfire/client/views/pfly/menu/megamenu/WorkbenchMegaMenuView.java]: @paulovmr to many logic on view. this should be placed on presenter right?

File: uberfire-workbench/uberfire-workbench-client-views-patternfly/src/main/java/org/uberfire/client/views/pfly/menu/megamenu/WorkbenchMegaMenuView.java [hunk 4]
```diff
@@ -0,0 +1,328 @@
...
+        menuItemByIdentifier = new HashMap<>();
+    }
+
+    @Override
+    public void addMenuItem(final String id,
+                            final String label,
+                            final String parentId,
+                            final Command command,
+                            final MenuPosition position) {
+        if (parentId == null || parentId.isEmpty()) {
```
Comment (ederign) [inline uberfire-workbench/uberfire-workbench-client-views-patternfly/src/main/java/org/uberfire/client/views/pfly/menu/megamenu/WorkbenchMegaMenuView.java]: same comment as above.

File: uberfire-workbench/uberfire-workbench-client-views-patternfly/src/main/java/org/uberfire/client/views/pfly/menu/megamenu/WorkbenchMegaMenuView.java [hunk 5]
```diff
@@ -0,0 +1,328 @@
...
+
+        menuItemByIdentifier.put(id,
+                                 groupMenuItemPresenter);
+
+        leftMenuItems.appendChild(groupMenuItemPresenter.getView().getElement());
+    }
+
+    @Override
+    public void selectMenuItem(final String id) {
+        // Do nothing
```
Comment (ederign) [inline uberfire-workbench/uberfire-workbench-client-views-patternfly/src/main/java/org/uberfire/client/views/pfly/menu/megamenu/WorkbenchMegaMenuView.java]: warning?

File: uberfire-workbench/uberfire-workbench-client-views-patternfly/src/main/java/org/uberfire/client/views/pfly/menu/megamenu/WorkbenchMegaMenuView.java [hunk 6]
```diff
@@ -0,0 +1,328 @@
...
+
+        menuItemByIdentifier.put(id,
+                                 groupContextMenuItemPresenter);
+
+        rightMenuItems.appendChild(groupContextMenuItemPresenter.getView().getElement());
+    }
+
+    @Override
+    public void clearContextMenu() {
+        // Do nothing
```
Comment (ederign) [inline uberfire-workbench/uberfire-workbench-client-views-patternfly/src/main/java/org/uberfire/client/views/pfly/menu/megamenu/WorkbenchMegaMenuView.java]: I wonder with so many ignored method if it's not a case of bad class hierarchy.

File: uberfire-workbench/uberfire-workbench-client-views-patternfly/src/main/java/org/uberfire/client/views/pfly/menu/megamenu/menuitem/GroupMenuItemView.java [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+    public void setLabel(final String label) {
+        title.setTextContent(label);
+    }
+
+    @Override
+    public void addItem(final IsElement item) {
+        items.appendChild(item.getElement());
+    }
+
+    @Override
```
Comment (ederign) [inline uberfire-workbench/uberfire-workbench-client-views-patternfly/src/main/java/org/uberfire/client/views/pfly/menu/megamenu/menuitem/GroupMenuItemView.java]: again :(

File: uberfire-workbench/uberfire-workbench-client-views-patternfly/src/main/java/org/uberfire/client/views/pfly/menu/megamenu/menuitem/IconMenuItemView.java [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+    Span icon;
+
+    @Override
+    public void init(final IconMenuItemPresenter presenter) {
+        this.presenter = presenter;
+    }
+
+    @Override
+    public void setIconClass(String iconClass) {
+        if (iconClass != null && !iconClass.isEmpty()) {
```
Comment (ederign) [inline uberfire-workbench/uberfire-workbench-client-views-patternfly/src/main/java/org/uberfire/client/views/pfly/menu/megamenu/menuitem/IconMenuItemView.java]: are you sure that  you need to parse multiple css classes and add one by one?

**Comment by paulovmr** [comment]
@ederign Thanks for the review! I think it's best if I create a new presenter for the mega menu, and don't use the existent one. This way all the issues you pointed out would be solved. Wdyt?

File: uberfire-workbench/uberfire-workbench-client-views-patternfly/src/main/java/org/uberfire/client/views/pfly/menu/megamenu/menuitem/IconMenuItemView.java (hunk 8 above)
Comment (paulovmr) [inline uberfire-workbench/uberfire-workbench-client-views-patternfly/src/main/java/org/uberfire/client/views/pfly/menu/megamenu/menuitem/IconMenuItemView.java]: Yes, it didn't work with only one add with multiple classes.

**Comment by paulovmr** [comment]
@ederign Hi, could you take another look? There are also new associated PRs, mentioned on the first comment in this PR. Thanks!

File: uberfire-workbench/uberfire-workbench-client-views-patternfly/src/main/java/org/uberfire/client/views/pfly/menu/SplashScreenMenuView.java [hunk 9]
```diff
@@ -41,6 +41,7 @@ public SplashScreenMenuView() {
         ensureDebugId("MenuSplashList-dropdown");
 
         anchor.addStyleName(Styles.DROPDOWN_TOGGLE);
+        anchor.addStyleName("nav-item-iconic");
```
Comment (ederign) [inline uberfire-workbench/uberfire-workbench-client-views-patternfly/src/main/java/org/uberfire/client/views/pfly/menu/SplashScreenMenuView.java]: constants?

File: uberfire-workbench/uberfire-workbench-client-views-patternfly/src/main/java/org/uberfire/client/views/pfly/menu/UserMenuViewImpl.java [hunk 10]
```diff
@@ -37,6 +37,7 @@
     @PostConstruct
     public void setup() {
         anchor.addStyleName(Styles.DROPDOWN_TOGGLE);
+        anchor.addStyleName("nav-item-iconic");
```
Comment (ederign) [inline uberfire-workbench/uberfire-workbench-client-views-patternfly/src/main/java/org/uberfire/client/views/pfly/menu/UserMenuViewImpl.java]: constants

File: uberfire-workbench/uberfire-workbench-client/src/main/java/org/uberfire/client/mvp/PerspectiveManagerImpl.java [hunk 11]
```diff
@@ -111,6 +118,30 @@ public void removePerspectiveStates(final Command doWhenFinished) {
         wbServices.removePerspectiveStates(doWhenFinished);
     }
 
+    @Override
+    public PerspectiveActivity getDefaultPerspectiveActivity() {
```
Comment (ederign) [inline uberfire-workbench/uberfire-workbench-client/src/main/java/org/uberfire/client/mvp/PerspectiveManagerImpl.java]: getIdentifier.

File: uberfire-workbench/uberfire-workbench-client/src/main/java/org/uberfire/client/workbench/widgets/menu/base/WorkbenchBaseMenuUtils.java [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+ * limitations under the License.
+ */
+
+package org.uberfire.client.workbench.widgets.menu.base;
+
+import org.uberfire.workbench.model.menu.MenuItem;
+
+public class WorkbenchBaseMenuUtils {
+
+    public String getMenuItemId(final MenuItem menuItem) {
```
Comment (ederign) [inline uberfire-workbench/uberfire-workbench-client/src/main/java/org/uberfire/client/workbench/widgets/menu/base/WorkbenchBaseMenuUtils.java]: static method.

Comment (ederign) [inline uberfire-workbench/uberfire-workbench-client/src/main/java/org/uberfire/client/workbench/widgets/menu/base/WorkbenchBaseMenuUtils.java]: (just suggestion)

File: uberfire-workbench/uberfire-workbench-client/src/main/java/org/uberfire/client/workbench/widgets/menu/megamenu/visitor/WorkbenchMegaMenuVisitor.java [hunk 13]
```diff
@@ -0,0 +1,135 @@
...
+    @Override
+    public void visit(final MenuItemPerspective menuItemPerspective) {
+        final String id = menuItemPerspective.getPlaceRequest().getIdentifier();
+        presenter.addMenuItem(id,
+                              menuItemPerspective.getCaption(),
+                              parentId,
+                              new Command() {
+                                  @Override
+                                  public void execute() {
+                                      IOC.getBeanManager().lookupBean(PlaceManager.class).getInstance().goTo(menuItemPerspective.getPlaceRequest());
```
Comment (ederign) [inline uberfire-workbench/uberfire-workbench-client/src/main/java/org/uberfire/client/workbench/widgets/menu/megamenu/visitor/WorkbenchMegaMenuVisitor.java]: leak of instances. if the old menu also does that, please uptade it.

**Comment by ederign** [comment]
Also, please check on home button if user has access (if (hasAccessToPerspective("changeme")) {)

**Comment by paulovmr** [comment]
@ederign All fixed!

File: uberfire-workbench/uberfire-workbench-client-views-patternfly/src/main/java/org/uberfire/client/views/pfly/menu/megamenu/menuitem/IconMenuItemView.java [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+            final String[] classes = iconClass.split(" ");
```
Comment (paulovmr) [inline uberfire-workbench/uberfire-workbench-client-views-patternfly/src/main/java/org/uberfire/client/views/pfly/menu/megamenu/menuitem/IconMenuItemView.java]: Hi @tiagodolphine , thanks, I will change this in the next PR ;)

**Pull request merged**